sdks/python/apache_beam/runners/interactive/display_graph.py

```diff
@@ -0,0 +1,232 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Helper to render pipeline graph in IPython when running interactively.
+
+This module is experimental. No backwards-compatibility guarantees.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import collections
+import re
+import threading
+
+import graphviz
+
+
+def nice_str(o):
+  s = repr(o)
+  s = s.replace('"', "'")
+  s = s.replace('\\', '|')
+  s = re.sub(r'[^\x20-\x7F]', ' ', s)
+  assert '"' not in s
+  if len(s) > 35:
+    s = s[:35] + '...'
+  return s
+
+
+def format_sample(contents, count=1000):
+  contents = list(contents)
+  elems = ', '.join([nice_str(o) for o in contents[:count]])
+  if len(contents) > count:
+    elems += ', ...'
+  assert '"' not in elems
+  return '{%s}' % elems
+
+
+def construct_digraph(vertex_dict, edge_dict):
+  """Given vertices, edges and their attributes, construct graphviz.Digraph.
+
+  Args:
+    vertex_dict: (dict(str->(dict(str->str))) dict mapping vertex
+        names to their attributes.
+    edge_dict: (dict(str, str)->(dict(str -> str))) dict mapping edge names to
+        their attributes.
+  Returns:
+    (graphviz.Digraph)
+  """
+  digraph = graphviz.Digraph()
+  for vertex, vertex_attributes in vertex_dict.items():
+    digraph.node(vertex, **vertex_attributes)
+  for edge, edge_attributes in edge_dict.items():
+    digraph.edge(edge[0], edge[1], **edge_attributes)
+  return digraph
+
+
+class PipelineGraph(object):
+  """Creates a DOT graph representation of a pipeline. Thread-safe."""
+
+  def __init__(self,
+               pipeline_proto,
+               required_transforms=None,
+               referenced_pcollections=None,
+               cached_pcollections=None,
+               pcollection_stats=None):
+    """Constructor of PipelineGraph.
+
+    All fields except for pipeline_proto should be left as None unless used by
```

Comment:
Rather than taking interactive-specific parameters here, take a map of {pcollections,ptransforms} -> extra attributes. The interactive runner would then populate these according to its notion of cached, etc. but other runners could highlight different data.